src/tools/perf/perftest_daemon.c

```diff
@@ -0,0 +1,511 @@
+{
+    ucp_perf_daemon_context_t *ctx = user_data;
+
+    ucp_perf_daemon_send_am_eager_msg(ctx->client_ep,
+                                      UCP_PERF_DAEMON_AM_ID_RECV_ACK);
+    ucp_request_free(request);
+}
+
+static void ucp_perf_daemon_ep_create(ucp_perf_daemon_context_t *ctx)
+{
+    ucp_ep_params_t ep_params;
+    ucs_status_t status;
+
+    ep_params.field_mask       = UCP_EP_PARAM_FIELD_FLAGS       |
+                                 UCP_EP_PARAM_FIELD_SOCK_ADDR   |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLER |
+                                 UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode         = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.err_handler.cb   = ucp_perf_daemon_err_cb;
+    ep_params.err_handler.arg  = ctx;
+    ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.sockaddr.addr    = (struct sockaddr*)&ctx->peer_address;
+    ep_params.sockaddr.addrlen = sizeof(ctx->peer_address);
+
+    status = ucp_ep_create(ctx->worker, &ep_params, &ctx->peer_ep);
+    if (status != UCS_OK) {
+        ucs_error("daemon failed to create an endpoint: %s",
+                  ucs_status_string(status));
+    }
+
+    ucp_perf_daemon_send_am_eager_msg(ctx->peer_ep,
+                                      UCP_PERF_DAEMON_AM_ID_PEER_INIT);
+}
+
+static void ucp_perf_daemon_check_am_msg(const ucp_am_recv_param_t *param,
+                                         size_t header_length,
+                                         int check_reply_ep)
+{
+    ucs_assert(!(param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV));
+    ucs_assertv(header_length == 0, "header_length %zu", header_length);
+
+    if (check_reply_ep) {
+        ucs_assert(param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP);
+    }
+}
+
+static ucs_status_t
+ucp_perf_daemon_init_handler(void *arg, const void *header,
+                             size_t header_length, void *data, size_t length,
+                             const ucp_am_recv_param_t *param)
+{
+    ucp_perf_daemon_context_t *ctx = arg;
+    void *p                        = data;
+    ucs_status_t status;
+    uint16_t address_length;
+    uint16_t send_memh_buf_size;
+    ucp_mem_attr_t attr;
+
+    ucp_perf_daemon_check_am_msg(param, header_length, 1);
+
+    if (ctx->peer_ep != NULL) {
+        ucs_fatal("duplicate daemon init req");
+        goto out;
+    }
+
+    ctx->client_ep = param->reply_ep;
+
+    address_length = *ucs_serialize_next(&p, uint16_t);
+
+    if (address_length != 0) {
+        memcpy(&ctx->peer_address,
+               ucs_serialize_next_raw(&p, void, address_length),
+               address_length);
+        ucp_perf_daemon_ep_create(ctx);
+    }
+
+    send_memh_buf_size = *ucs_serialize_next(&p, uint16_t);
+    ctx->send_memh     = ucp_perf_daemon_memh_import(ctx, p);
+    ucs_serialize_next_raw(&p, void, send_memh_buf_size);
+    ucs_serialize_next(&p, uint16_t); /*recv_memh_buf_size*/
```

Comment:
My understanding is that this is a generic serialization of LV pairs (Length-Value). That's right, length of the last item in the packet is a kind of redundant, because there is nothing behind it. But I would keep it just for consistency and uniformity reasons, so that all the items are LV.
It's also more extensible, as we can easily add more LVs without changing existing format